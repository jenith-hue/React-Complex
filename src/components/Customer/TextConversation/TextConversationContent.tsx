import {
  makeStyles,
  RACCOLOR,
  RACSelect,
  RACTextConversationMessage,
  RACTextMessageInput,
  Typography,
} from '@rentacenter/racstrap';
import React, { useEffect, useState, useRef } from 'react';
import clsx from 'clsx';
import {
  useTextConversation,
  useTextConversationActions,
} from '../../../context/TextConversationProvider/TextConversationProvider';
import { TextTemplateModal } from '../TextTemplate/TextTemplateModal';
import { ApiStateWrapper } from '../../../common/ApiStateWrapper/ApiStateWrapper';
import { formatPhoneNumber } from '../../../utils/utils';
import { useStoreDetails } from '../../../context/Store/StoreProvider';
import { useUserPermissions } from '../../../context/permission/PermissionsProvider';
import { useUserStateContext } from '../../../context/user/user-contexts';
import { FRCStoreType, RentACenter } from '../../../constants/constants';

export interface TextConversationContentProps {
  onComponentDidMount: () => void;
}

const useStyles = makeStyles((theme: any) => ({
  root: {
    padding: '1rem 0.5rem 1rem 0.5rem',
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  header: {
    display: 'flex',
    width: '100%',
    // eslint-disable-next-line sonarjs/no-duplicate-string
    justifyContent: 'flex-start',
  },
  body: {
    width: '100%',
    overflowY: 'auto',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-end',
  },
  noRecord: {
    display: 'flex',
    justifyContent: 'center !important',
  },
  headerItem: {
    display: 'flex',
    alignItems: 'flex-start',
    flexDirection: 'column',
    flex: 1,
  },
  headerItemValue: {
    color: `${RACCOLOR.INDEPENDENCE} !important`,
    whiteSpace: 'nowrap',
  },
  centered: {
    alignItems: 'center',
  },
  messageDate: {
    fontSize: `${theme.typography.pxToRem(12)} !important`,
    color: `${RACCOLOR.ARGENT} !important`,
    marginTop: '1rem',
    marginBottom: '1rem',
  },
  leftMargin: {
    marginLeft: '2rem',
  },
}));
export const TextConversationContent = ({
  onComponentDidMount,
}: TextConversationContentProps) => {
  const classes = useStyles();
  const ref = useRef(null);
  const {
    phoneOptions,
    selectedPhoneOption,
    fullName,
    messages,
    isLoading,
    hasApiError,
    preferredLanguage,
    isCommunicationAllowed,
  } = useTextConversation();

  const { onPhoneOptionChanged, sendMessage } = useTextConversationActions();
  const [openTextTemplateModal, setOpenTextTemplateModal] = useState(false);
  const { store } = useStoreDetails();
  const { user } = useUserStateContext();
  const CoworkerName = `${user?.firstName} ${user?.lastName
    .charAt(0)
    .toUpperCase()}`;
  const companyName =
    store?.storeType === FRCStoreType
      ? `${CoworkerName} ${store?.storeNumber}`
      : RentACenter;
  const { hasAccessToTextTemplate, hasAccessToFreeFormText } =
    useUserPermissions();
  const DEFAULT_PRE_INPUT_TEXT = 'Templates';
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(onComponentDidMount, []);

  useEffect(() => {
    if (ref?.current) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      ref?.current?.scroll({
        top: Number.MAX_SAFE_INTEGER,
        left: 0,
        behavior: 'smooth',
      });
    }
  }, [messages, isLoading]);

  return (
    <div className={classes.root}>
      <div
        className={clsx(
          classes.body,
          isLoading || !messages.length || hasApiError ? classes.noRecord : null
        )}
        ref={ref}
      >
        <ApiStateWrapper
          loading={isLoading}
          hasApiError={hasApiError}
          response={messages}
          successContent={
            <>
              {messages.map(
                ({
                  textMessageId,
                  isSentByRAC,
                  message,
                  messageDate,
                  initials,
                  tooltipTitle,
                  tooltipSubTitle,
                }) => (
                  <div key={textMessageId}>
                    <Typography
                      align="center"
                      variant="body1"
                      className={classes.messageDate}
                    >
                      {messageDate}
                    </Typography>
                    <RACTextConversationMessage
                      key={textMessageId}
                      isOwnMessage={isSentByRAC}
                      avatarText={initials}
                      message={message}
                      tooltipTitle={tooltipTitle}
                      tooltipSubtitle={tooltipSubTitle}
                    />
                  </div>
                )
              )}
            </>
          }
        />
      </div>
      <div className={classes.header}>
        <div className={classes.headerItem}>
          <Typography align="left" variant="h5">
            Name
          </Typography>
          <Typography
            className={classes.headerItemValue}
            variant="body1"
            align="left"
          >
            {fullName}
          </Typography>
        </div>
        <div className={clsx(classes.headerItem, classes.leftMargin)}>
          <Typography align="left" variant="h5">
            Phone
          </Typography>
          <RACSelect
            isDisabled={!isCommunicationAllowed || !phoneOptions?.length}
            defaultValue={selectedPhoneOption?.value}
            options={phoneOptions.map((phone) => ({
              ...phone,
              label: formatPhoneNumber(phone.label, !isCommunicationAllowed),
            }))}
            onChange={(e: React.ChangeEvent<{ value: any }>) => {
              onPhoneOptionChanged(e.target.value);
            }}
          />
        </div>

        <div className={clsx(classes.headerItem, classes.centered)}>
          <Typography align="center" variant="h5">
            Type
          </Typography>
          <Typography
            align="center"
            className={classes.headerItemValue}
            variant="body1"
          >
            {selectedPhoneOption?.phoneType}
          </Typography>
        </div>
        <div className={classes.headerItem}>
          <Typography align="center" variant="h5">
            Language
          </Typography>
          <Typography
            align="center"
            className={classes.headerItemValue}
            variant="body1"
          >
            {preferredLanguage}
          </Typography>
        </div>
      </div>
      {hasAccessToTextTemplate && hasAccessToFreeFormText && (
        <div>
          <Typography variant="h5">Send Text To:</Typography>
          <Typography variant="body1">
            Choose Text Template or Free Form Text
          </Typography>
        </div>
      )}
      <TextTemplateModal
        open={openTextTemplateModal}
        onClose={setOpenTextTemplateModal}
        preferredLanguageDesc={preferredLanguage}
        isCommunicationAllowed={isCommunicationAllowed}
      />
      <RACTextMessageInput
        onSend={(e) => {
          sendMessage(`${e} - ${companyName}`);
        }}
        onPreInputTextClicked={() =>
          hasAccessToTextTemplate && setOpenTextTemplateModal(true)
        }
        isCommunicationAllowed={
          isCommunicationAllowed &&
          hasAccessToFreeFormText &&
          phoneOptions?.length > 0
        }
        preInputText={
          hasAccessToTextTemplate && phoneOptions?.length > 0
            ? DEFAULT_PRE_INPUT_TEXT
            : ' '
        }
      />
    </div>
  );
};
