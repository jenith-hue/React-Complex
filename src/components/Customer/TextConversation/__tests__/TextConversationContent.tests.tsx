/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { TextConversationContent } from '../TextConversationContent';

interface PropWithChildren {
  children: JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACTextMessageInput: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTextMessageInput" {...rest}>
      {children}
    </div>
  ),
  RACTextConversationMessage: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTextConversationMessage" {...rest}>
      {children}
    </div>
  ),
  RACSelect: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACSelect" {...rest}>
      {children}
    </div>
  ),
  RACModal: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACModal" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({
    root: 'root',
    header: 'header',
    body: 'body',
    footer: 'footer',
    headerItem: 'headerItem',
    headerItemValue: 'headerItemValue',
    message: 'message',
    mirror: 'mirror',
    avatarContainer: 'avatarContainer',
    contentContainer: 'contentContainer',
  }),
  RACCOLOR: {
    INDEPENDENCE: 'INDEPENDENCE',
  },
}));

jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetails: () => ({
      customerDetails: { firstName: '', lastName: '', phones: [] },
      coCustomerDetails: { firstName: '', lastName: '', phones: [] },
    }),
  })
);

jest.mock(
  '../../../../context/TextConversationProvider/TextConversationProvider',
  () => ({
    useTextConversationActions: () => ({
      onPhoneOptionChanged: jest.fn(),
      sendMessage: jest.fn(),
    }),
    useCustomerDetails: () => ({
      customerDetails: { firstName: '', lastName: '', phones: [] },
      coCustomerDetails: { firstName: '', lastName: '', phones: [] },
    }),
    useTextConversation: () => ({
      isCommunicationAllowed: true,
      phoneOptions: [
        {
          label: 'label1',
          value: 'value1',
        },
        {
          label: 'label2',
          value: 'value2',
        },
      ],
      selectedPhoneOption: {
        label: 'label1',
        value: 'value1',
      },
      fullName: 'Full name',
      phoneType: 'CELL',
      messages: [
        {
          textMessageId: 'textMessageId1',
          isSentByRAC: true,
          phoneNumber: '555-555-555',
          messageDate: '10/10/2010 11:00',
          initials: 'AB',
          fullName: 'Abraham Bahamas',
          message: 'Hi',
        },
        {
          textMessageId: 'textMessageId2',
          isSentByRAC: false,
          phoneNumber: '33-2222-551115',
          messageDate: '10/10/2010 11:50',
          initials: 'JD',
          fullName: 'Jane Doe',
          message: 'Hi 2',
        },
      ],
    }),
    useUserPermissions: () => ({
      hasAccessToTextTemplate: true,
      hasAccessToFreeFormText: true,
    }),
  })
);

describe('TextConversationContent', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <TextConversationContent
        onComponentDidMount={() => {
          // noop
        }}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('calls onComponentDidMount after first mount ', () => {
    const mockOnComponentDidMount = jest.fn();

    renderer.act(() => {
      renderer.create(
        <TextConversationContent
          onComponentDidMount={mockOnComponentDidMount}
        />
      );
    });

    expect(mockOnComponentDidMount).toHaveBeenCalledTimes(1);
  });
});
