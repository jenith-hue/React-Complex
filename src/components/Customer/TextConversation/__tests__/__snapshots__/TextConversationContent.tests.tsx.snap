// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextConversationContent renders correctly 1`] = `
<div
  className="root"
>
  <div
    className="body"
  >
    <div>
      <div
        align="center"
        data-testid="Typography"
        variant="body1"
      >
        10/10/2010 11:00
      </div>
      <div
        avatarText="AB"
        data-testid="RACTextConversationMessage"
        isOwnMessage={true}
        message="Hi"
      />
    </div>
    <div>
      <div
        align="center"
        data-testid="Typography"
        variant="body1"
      >
        10/10/2010 11:50
      </div>
      <div
        avatarText="JD"
        data-testid="RACTextConversationMessage"
        isOwnMessage={false}
        message="Hi 2"
      />
    </div>
  </div>
  <div
    className="header"
  >
    <div
      className="headerItem"
    >
      <div
        align="left"
        data-testid="Typography"
        variant="h5"
      >
        Name
      </div>
      <div
        align="left"
        className="headerItemValue"
        data-testid="Typography"
        variant="body1"
      >
        Full name
      </div>
    </div>
    <div
      className="headerItem"
    >
      <div
        align="left"
        data-testid="Typography"
        variant="h5"
      >
        Phone
      </div>
      <div
        data-testid="RACSelect"
        defaultValue="value1"
        isDisabled={false}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "(lab) el1-",
              "value": "value1",
            },
            Object {
              "label": "(lab) el2-",
              "value": "value2",
            },
          ]
        }
      />
    </div>
    <div
      className="headerItem"
    >
      <div
        align="center"
        data-testid="Typography"
        variant="h5"
      >
        Type
      </div>
      <div
        align="center"
        className="headerItemValue"
        data-testid="Typography"
        variant="body1"
      />
    </div>
    <div
      className="headerItem"
    >
      <div
        align="center"
        data-testid="Typography"
        variant="h5"
      >
        Language
      </div>
      <div
        align="center"
        className="headerItemValue"
        data-testid="Typography"
        variant="body1"
      />
    </div>
  </div>
  <div
    buttons={
      <React.Fragment>
        <UNDEFINED
          color="secondary"
          onClick={[Function]}
          variant="outlined"
        >
          Cancel
        </UNDEFINED>
        <UNDEFINED
          color="primary"
          disabled={false}
          onClick={[Function]}
          variant="contained"
        >
          Send
        </UNDEFINED>
      </React.Fragment>
    }
    classes={
      Object {
        "dialog": undefined,
        "dialogContent": undefined,
      }
    }
    content={
      <TextTemplateContent
        onCategoryChange={[Function]}
        onQuickPhaseChange={[Function]}
      />
    }
    data-testid="RACModal"
    isOpen={false}
    maxWidth="md"
    onClose={[Function]}
    title="Text Templates"
  />
  <div
    data-testid="RACTextMessageInput"
    isCommunicationAllowed={false}
    onPreInputTextClicked={[Function]}
    onSend={[Function]}
    preInputText=" "
  />
</div>
`;
